### PR TITLE
[SPARK-26744][SQL]Support schema validation in FileDataSourceV2 framework

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

The file source has a schema validation feature, which validates 2 schemas:
1. the user-specified schema when reading.
2. the schema of input data when writing.

If a file source doesn't support the schema, we can fail the query earlier.

This PR is to implement the same feature  in the `FileDataSourceV2` framework. Comparing to `FileFormat`, `FileDataSourceV2` has multiple layers. The API is added in two places:
1. Read path: the table schema is determined in `TableProvider.getTable`. The actual read schema can be a subset of the table schema.  This PR proposes to validate the actual read schema in  `FileScan`.
2.  Write path: validate the actual output schema in `FileWriteBuilder`.

## How was this patch tested?

Unit test

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 1]
```diff
@@ -490,9 +490,6 @@ case class DataSource(
       outputColumnNames: Seq[String],
       physicalPlan: SparkPlan): BaseRelation = {
     val outputColumns = DataWritingCommand.logicalPlanOutputWithNames(data, outputColumnNames)
-    if (outputColumns.map(_.dataType).exists(_.isInstanceOf[CalendarIntervalType])) {
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: I think the validation here is duplicated with the check `supportDataType`. So I remove it. I can revert it if someone has a good reason for keeping it.

**Comment by gengliangwang** [comment]
@cloud-fan @dongjoon-hyun @HyukjinKwon

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala [hunk 2]
```diff
@@ -18,22 +18,29 @@ package org.apache.spark.sql.execution.datasources.v2
...
    */
   def isSplitable(path: Path): Boolean = {
     false
   }
 
+  /**
+   * Returns whether this format supports the given [[DataType]] in write path.
+   * By default all data types are supported.
+   */
+  def supportDataType(dataType: DataType): Boolean = true
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala]: Hi, @gengliangwang and @cloud-fan .
In DSv2, I guess it would be more natural to have a Java interface for this validation API. How do you think about that?

cc @rdblue since this is DSv2.

**Comment by rdblue** [comment]
@gengliangwang, can you be more clear about what you are proposing to add to DSv2?

I don't think that simply porting an API from v1 isn't sufficient justification to add to v2, because v1 has so many problems. I'd like to see a description on the JIRA issue that states exactly what is added and how it changes behavior.

Until then, please consider this a -1.

**Comment by gengliangwang** [comment]
@rdblue @dongjoon-hyun Sorry about the confuse. This PR is for an internal API in file source V2 only (abstract class `FileWriteBuilder` and `FileScan`).

I have updated the PR description.

**Comment by AsmaZgo** [comment]
hello, 
I'm sorry my question is not directly related to the context. 
In my use case, I need to access to all the logical plans generated by the optimizer (not just the optimal one). how can I do that with spark SQL?
 thank you very much

**Comment by rdblue** [comment]
@gengliangwang, why are you proposing to add this API that applies only to internal sources? Why not design this to work with all sources?

I think you also need to be more clear about what you're trying to commit. What does this do? It sounds like it probably validates that a file format can stores a type. For example, can ORC support DECIMAL(44, 6)? That is generally useful. Why should it be a side API for internal sources?

In short:
* Please be clear in your description about what this commit does. What exactly does the validation do?
* Please give a reason why it should apply only to internal sources given that a goal of the DSv2 API is to avoid special cases for internal sources.

**Comment by gengliangwang** [comment]
@AsmaZgo I think you can see how the plan changes in log by setting `spark.sql.optimizer.planChangeLog.level`. For further questions, please send email to dev@spark.apache.org

**Comment by gengliangwang** [comment]
@rdblue You are right. Thanks for the suggestion. I will try making it a DSV2 API.

**Comment by cloud-fan** [comment]
@rdblue @gengliangwang I don't think this needs an API change. This is just a schema validation feature, which can be done in any DS v2 sources. Schema validation needs to know the user-specified schema when reading, or the schema of input data when writing, which are both available in the current ds v2 APIs. It looks to me this PR just re-implements the schema validation feature in file source v2 framework.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by rdblue** [comment]
@cloud-fan, should this validation be done for all sources, or just for file sources for some reason?

I would also like to know *exactly what validation is proposed in this PR*. It hasn't been written up and I think that a summary of the changes that are proposed is required before we commit changes.

**Comment by cloud-fan** [comment]
every source can do schema validation if needed, the DS v2 API already allows you to do so. To save duplicated code, this PR proposes to add the entry point of schema validation in the base class of file source.

@gengliangwang Can you post the details of the schema validation for these file sources?

**Comment by cloud-fan** [comment]
@rdblue you can treat this PR as implementing schema validation for file sources. We only do it for file sources because for now they are the only DS v2 implementation we have in Spark.

**Comment by gengliangwang** [comment]
The supported data types of file sources:

- text: String
- json: AtomicType/StructType/ArrayType/MapType/UDT/NULLType
- CSV: AtomicType/UserDefinedType
- ORC: AtomicType/StructType/ArrayType/MapType/udt
- ...

For details please read https://github.com/apache/spark/pull/21667

This is very simple abstraction.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcScan.scala [hunk 3]
```diff
@@ -40,4 +40,10 @@ case class OrcScan(
     OrcPartitionReaderFactory(sparkSession.sessionState.conf, broadcastedConf,
       dataSchema, fileIndex.partitionSchema, readSchema)
   }
+
+  override def supportDataType(dataType: DataType): Boolean = {
+    OrcDataSourceV2.supportDataType(dataType)
+  }
+
+  override def toString: String = "ORC"
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcScan.scala]: There is much more useful information in this class than just the file format name. This should use a `formatName` method instead so that `toString` can be used to show the object itself when debugging or logging.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileWriteBuilder.scala [hunk 4]
```diff
@@ -104,12 +104,24 @@ abstract class FileWriteBuilder(options: DataSourceOptions)
       options: Map[String, String],
       dataSchema: StructType): OutputWriterFactory
 
+  /**
+   * Returns whether this format supports the given [[DataType]] in write path.
+   * By default all data types are supported.
+   */
+  def supportDataType(dataType: DataType): Boolean = true
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileWriteBuilder.scala]: This name is awkward. Similar methods or traits use "supports" instead of "support". I think this should as well.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileWriteBuilder.scala [hunk 5]
```diff
@@ -104,12 +104,24 @@ abstract class FileWriteBuilder(options: DataSourceOptions)
...
   private def validateInputs(): Unit = {
     assert(schema != null, "Missing input data schema")
     assert(queryId != null, "Missing query ID")
     assert(mode != null, "Missing save mode")
     assert(options.paths().length == 1)
     DataSource.validateSchema(schema)
+    schema.foreach { field =>
+      if (!supportDataType(field.dataType)) {
+        throw new AnalysisException(
+          s"$this data source does not support ${field.dataType.catalogString} data type.")
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileWriteBuilder.scala]: Should use `formatName` instead of `$this` (toString)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala [hunk 6]
```diff
@@ -18,22 +18,29 @@ package org.apache.spark.sql.execution.datasources.v2
...
 import org.apache.spark.sql.execution.datasources._
 import org.apache.spark.sql.sources.v2.reader.{Batch, InputPartition, Scan}
-import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.types.{DataType, StructType}
 
 abstract class FileScan(
     sparkSession: SparkSession,
-    fileIndex: PartitioningAwareFileIndex) extends Scan with Batch {
+    fileIndex: PartitioningAwareFileIndex,
+    readSchema: StructType) extends Scan with Batch {
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala]: @gengliangwang, why validate the read schema here in FileScan instead of in the scan builder?

**Comment by rdblue** [comment]
@cloud-fan, thanks for the clarifications, particularly the updated description.

I don't think we need to add type validation to v2 yet. This is something that could be done in that API, but I'm not sure that it is a good idea to standardize it because that would make assumptions about why types are not supported. For example, using a capability-based API for types like `int` or `struct` sounds reasonable, but doesn't work for a delimited format that can support some nesting, but not arbitrarily deep nesting.

@gengliangwang, I flagged a couple of review items to address. In addition, I would recommend taking more care when answering questions. It is concerning that I wasn't able to get a concise answer from you about what you're proposing to change. Of course I can go down a rabbit-hole of trying to find out what your intent is by reading code and other pull requests. But it is much easier for everyone if you clearly state what you're proposing and why.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileWriteBuilder.scala (hunk 4 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileWriteBuilder.scala]: I proposed to change the name as "supports..." in previous PR: https://github.com/apache/spark/pull/23639

See the opposed comment here:  https://github.com/apache/spark/pull/23639#pullrequestreview-196362643

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala (hunk 6 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala]: In the PR description:
> the table schema is determined in `TableProvider.getTable`. The actual read schema can be a subset of the table schema.  This PR proposes to validate the actual read schema in  `FileScan`

**Comment by gengliangwang** [comment]
@rdblue For example, when Spark try writing data contains Array type column to CSV source:
1. without the validation: there will be exceptions on execution tasks. 
2. with the validation: there will be exception before launching tasks. Also, the error message is more user-friendly.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: is this method only called by file sources?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 7]
```diff
@@ -527,10 +524,6 @@ case class DataSource(
    * Returns a logical plan to write the given [[LogicalPlan]] out to this [[DataSource]].
    */
   def planForWriting(mode: SaveMode, data: LogicalPlan): LogicalPlan = {
-    if (data.schema.map(_.dataType).exists(_.isInstanceOf[CalendarIntervalType])) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala]: per discussion, this is an implementation in file source, not an API. It's internal so we don't need java here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileWriteBuilder.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileWriteBuilder.scala]: This one is different. We are in a new class, and there is no `supportXXX` method in this class that we need to follow.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileWriteBuilder.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileWriteBuilder.scala]: +1

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcScan.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcScan.scala]: +1

**Comment by gengliangwang** [comment]
retest this please.

**Comment by rdblue** [comment]
+1

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 8]
```diff
@@ -530,7 +527,6 @@ case class DataSource(
     if (data.schema.map(_.dataType).exists(_.isInstanceOf[CalendarIntervalType])) {
       throw new AnalysisException("Cannot save interval data type into external storage.")
     }
-
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: unnecessary change

**Comment by cloud-fan** [comment]
thanks, merging to master!
